setup.py

```diff
@@ -83,76 +218,430 @@ def findRequirements(repositoryDir):
+    "-outputtuple",
+    "-castmode",
+    "-nosafecstrings",
+    "-w402", #TODO silence warnings
+    "-w503",
+    "-w511",
+    "-w302",
+    "-w362",
+    "-w312",
+    "-w389",
+    "-DSWIG_PYTHON_LEGACY_BOOL",
+    "-I" + swigDir + "/python",
+    "-I" + swigDir]
+  for define in commonDefines:
+    item = "-D" + define[0]
+    if define[1]:
+      item += define[0] + "=" + define[1]
+    swigFlags.append(item)
+  for includeDir in commonIncludeDirs:
+    item = "-I" + includeDir
+    swigFlags.append(item)
+
+  wrapAlgorithms = generateSwigWrap(swigExecutable,
+                                    swigFlags,
+                                    "nupic/bindings/algorithms.i")
+  libModuleAlgorithms = Extension(
+    "nupic.bindings._algorithms",
+    extra_compile_args=commonCompileFlags,
+    define_macros=commonDefines,
+    extra_link_args=commonLinkFlags,
+    include_dirs=commonIncludeDirs,
+    libraries=commonLibraries,
+    sources=pythonSupportSources + [wrapAlgorithms],
+    extra_objects=commonObjects)
+  extensions.append(libModuleAlgorithms)
+
+  wrapEngineInternal = generateSwigWrap(swigExecutable,
+                                        swigFlags,
+                                        "nupic/bindings/engine_internal.i")
+  libModuleEngineInternal = Extension(
+    "nupic.bindings._engine_internal",
+    extra_compile_args=commonCompileFlags,
+    define_macros=commonDefines,
+    extra_link_args=commonLinkFlags,
+    include_dirs=commonIncludeDirs,
+    libraries=commonLibraries,
+    sources=pythonSupportSources + [wrapEngineInternal],
+    extra_objects=commonObjects)
+  extensions.append(libModuleEngineInternal)
+
+  wrapMath = generateSwigWrap(swigExecutable,
+                              swigFlags,
+                              "nupic/bindings/math.i")
+  libModuleMath = Extension(
+    "nupic.bindings._math",
+    extra_compile_args=commonCompileFlags,
+    define_macros=commonDefines,
+    extra_link_args=commonLinkFlags,
+    include_dirs=commonIncludeDirs,
+    libraries=commonLibraries,
+    sources=pythonSupportSources + [wrapMath,
+                                    "nupic/bindings/PySparseTensor.cpp"],
+    extra_objects=commonObjects)
+  extensions.append(libModuleMath)
+
+  return extensions
+
+
+
+def getCommandLineOption(name, options):
+  if name in options:
+    return options[name]
+
+
+
+def prepareNupicCore(options, platform, bitness):
+
+  nupicCoreReleaseDir = getCommandLineOption("nupic-core-dir", options)
+  nupicCoreSourceDir = None
+  fetchNupicCore = True
+
+  if nupicCoreReleaseDir:
+    # User specified that they have their own nupic.core
+    fetchNupicCore = False
+  else:
+    nupicCoreReleaseDir, nupicCoreSourceDir = getDefaultNupicCoreDirectories()
+
+  nupicCoreCommitish = extractNupicCoreTarget()
+
+  if fetchNupicCore:
+    # User has not specified 'nupic.core' location, so we'll download the
+    # binaries.
+
+    nupicCoreRemoteUrl = (NUPIC_CORE_BUCKET + "/nupic_core-"
```

Comment:
nit: not necessary for merge but string formatting might make this easier to understand